unicorn/src/unicorn/zookeeper.cpp

```diff
@@ -96,27 +97,42 @@ zookeeper_t::put(
     const value_t& value,
     version_t version
 ) {
+    auto scope = std::make_shared<zk_scope_t>();
     if (version < 0) {
-        auto ec = make_error_code(cocaine::error::version_not_allowed);
-        auto future = make_exceptional_future<response::put>(std::move(ec));
-        callback(std::move(future));
-        return nullptr;
+        //TODO: use move capture when it is available
+        executor->spawn(std::bind([](callback::put& cb){
+            auto ec = make_error_code(cocaine::error::version_not_allowed);
+            auto future = make_exceptional_future<response::put>(std::move(ec));
+            cb(std::move(future));
+        }, std::move(callback)));
+
+        return scope;
     }
-    auto scope = std::make_shared<zk_scope_t>();
+
     std::shared_ptr<logging::logger_t> logger = context.log(cocaine::format("unicorn/{}", name), {
         {"method", "put"},
         {"path", path},
         {"version", version},
         {"command_id", rand()}});
+
     auto& handler = scope->handler_scope.get_handler<put_action_t>(
         context_t({logger, zk}),
         std::move(callback),
         path,
         value,
         std::move(version)
     );
+
     COCAINE_LOG_DEBUG(logger, "start processing");
-    zk.put(handler.path, handler.encoded_value, handler.version, handler);
+    try {
+        zk.put(handler.path, handler.encoded_value, handler.version, handler);
+    } catch (...) {
```

Comment:
try_run([]{
        zk.put(handler.path, handler.encoded_value, handler.version, handler);
    }, handler.callback, *executor);

Would be better?